### ISSUE TITLE
[fix proposal] Super slow with multiple descendant tokens

### ISSUE THREAD

**Issue by dbuezas** [fb55/css-select#93]
for reference: https://github.com/cheeriojs/cheerio/issues/1118

Since the descendant token has the option to "consume" an element or to skip to an ancestor, the time complexity explodes with selectors like `.row div div div div`. In these cases, for all elements that are not descendants of `.row` the SAME `descendant` function will be called many times on the same ancestors, resulting in exponential worst case time complexity (relative to number of descendant tokens in the selector).

I have two solutions that cache previous results and they solve the issue. The fastest one is invasive (adds an ID to the elements) the other solution uses a local Set to cache results so it leaves the element object untouched, but is twice as slow in node v9.2.0 (but not in v6.10.0)

`file: lib/general.js`

Slow solution:
This one does not mutate the element but requires node > 0.12.8 (or a polyfill for Set)
```javascript
descendant: function(next){
	var isFalseCache = new Set();

	return function descendant(elem){
		var found = false;

		while(!found && (elem = adapter.getParent(elem))){
			if(!isFalseCache.has(elem)){
				found = next(elem);
				if(!found){
					isFalseCache.add(elem);
				}
			}
		}

		return found;
	};
},
```

Fast (but mutates element by adding an Id to it) solution:
This one does not add any measurable delay to normal cases in my test.
```javascript
var elemId = 1;
...
descendant: function(next){
	var isFalseCache = [];

	return function descendant(elem){
		var found = false;

		while(!found && (elem = adapter.getParent(elem))){
			if(!isFalseCache[elem._id]){
				found = next(elem);
				if(!found){
					if(!elem._id) elem._id = elemId++;
					isFalseCache[elem._id] = true;
				}
			}
		}

		return found;
	};
},
```
(benchmarking done with around 3000 selectors across 30 pages)

If somebody could give me feedback I will make a PR.
Props to the devs btw, this lib is super clever!

**Comment by fb55** [comment]
This looks great, thanks for digging in! I would strongly prefer the set approach, node < 4 shouldn't be a concern. I take it the slowdown only exists in node 9, but not 6?

**Comment by dbuezas** [comment]
Oh great! Thanks for the feedback.
Will PR when I come back from my vacations. I'll try to provide the test case also.
We've been using this modification for two weeks and the occasional long run times seem to be gone for good.

**Comment by dbuezas** [comment]
Yes, that is correct, i didn't observe slowdowns in node 6 when using Sets

**Comment by fb55** [comment]
Fixed in #94!
